### ISSUE TITLE
POCL CPU performance vs native Intel OpenCL runtime

### ISSUE THREAD

**Issue by jsteube** [pocl/pocl#292]
Not really an issue, just fyi!

Using oclHashcat the native Intel OpenCL runtime is almost twice as fast executing the same code than pocl. Since you should be able to reproduce this easily I thought I'd just show you how to reproduce it locally and maybe you get an Idea how to improve it:
- Get latest oclHashcat from GIT
- Install Intel OpenCL driver (I'm using 14.2 but latest one should work, too)
- Install pocl
- Run the command: ./oclHashcat -b -m 1000 --opencl-device-types 1 --opencl-platforms 1
- Run the command: ./oclHashcat -b -m 1000 --opencl-device-types 1 --opencl-platforms 2

Rest should be self explaining. Here's how it looks for me:

Intel:

```
Device #1: Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz, 3999/15999 MB allocatable, 3500Mhz, 8MCU
Speed.Dev.#1.:  1019.7 MH/s
```

pocl:

```
Device #6: Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz, 15999/15999 MB allocatable, 3888Mhz, 8MCU
Speed.Dev.#6.:   522.7 MH/s
```

If I'd have to guess what's the problem here is that pocl doesn't make use of the 256 bit AVX2 registers and uses 128 bit only. But that's why I'm posting this here, you can reproduce locally and find out yourself.

I did the same test on my AMD FX-8120 CPU. The speed between the native AMD CPU OpenCL runtime (part of Catalyst) and pocl is pretty close to each other that's why the results for Intel is very interessting.

This was tested with LLVM 3.7 build from sources and latest and greatest pocl from GIT.

**Comment by franz** [comment]
> pocl doesn't make use of the 256 bit AVX2 registers and uses 128 bit only

Probably correct, i'd guess the LLVM autovectorizer is unable to widen whatever OpenCL vectors you use to 256bit vectors. If you use 256bit OpenCL vectors directly (eg float8, int8), it should use the 256bit registers of your CPU.

The reason the FX isn't any faster is probably because the shared FPU of FX can work on either one 256bit or two 128bit instructions at a time, so using 256bit avx only gains you speed if the program is not using all cores.

**Comment by illwieckz** [comment]
> I did the same test on my AMD FX-8120 CPU. The speed between the native AMD CPU OpenCL runtime (part of Catalyst) and pocl is pretty close to each other that's why the results for Intel is very interessting.

One interesting experiment would be to try the AMD OpenCL stack with your Intel CPU, so you will get three results for the same CPU, using the Intel stack, using the AMD stack, using the POCL stack. If POCL and AMD results are pretty close on an AMD CPU, will they be close together on an Intel CPU too? Is Intel twice faster than AMD too on their Intel CPU, or is POCL half slower than the two others?

I haven't tried oclHashcat but with another benchmark (luxmark), the AMD and Intel results are close on the same Intel CPU. Intel is a bit faster, but not twice. So I'm surprised than POCL is half the speed of Intel, but equivalent to AMD since my experience show Intel and AMD are close, I was naively expecting POCL being half the speed of AMD too if it's half the speed of Intel. I can't try POCL on my own side yet due to a bug.

**Comment by Kazhuu** [comment]
Hello! I'm trying to look into this issue but I'm not able to reproduce it. On both Intel OpenCL CPU Runtime and pocl I'm getting about the same results with following command:
```
hashcat -b -m1000
```
I'm changing between Intel and pocl using variable `OCL_ICD_VENDORS`. See `man libOpenCL.so`.

Intel:
```
OpenCL Platform #1: Intel(R) Corporation
========================================
Device #1: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz, 3974/15898 MB allocatable, 4MCU
Speed.#1.........:   490.1 MH/s (8.33ms) @ Accel:1024 Loops:1024 Thr:1 Vec:8
```
pocl:
```
OpenCL Platform #1: The pocl project
====================================
Device #1: pthread-Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz, 4096/13850 MB allocatable, 4MCU
Speed.#1.........:   446.9 MH/s (9.10ms) @ Accel:1024 Loops:1024 Thr:1 Vec:8
```
Version information:
* hashcat, v5.1.0
* pocl: v1.5
* Intel OpenCL CPU runtime: v18.1

What I'm wondering how you @jsteube got over 1000 MH/s with Intel. Maybe it's related to changes in newer hashcat versions and now Intel implementation is not able to optimize it anymore like it used to or something. Do you remember which hashcat version you were using for your testing?

**Comment by Kazhuu** [comment]
Thanks to @franz I actually noticed that I overlooked a tiny detail. My laptop CPU is 2 cores and 4 threads and @jsteube is 4 cores and 8 threads. I'll try to reproduce this on other hardware and I should get around 1000 MH/s as well.

**Comment by Kazhuu** [comment]
Before I'm done with this issue I'll post my findings here so far if someone finds them interesting. With the latest pocl v1.5, LLVM 10, Intel OpenCtraL runtime (v18.1) and hashcat (v5.1.0) this issue shouldn't exist anymore. I'm able to get following result with `hashcat -b -m 1000 --force`. Which is close compared to the Intel implementation.
```
OpenCL Platform #1: The pocl project
====================================
* Device #1: pthread-Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz, 8192/29945 MB allocatable, 8MCU
...
Speed.#1.........:   822.5 MH/s (9.67ms) @ Accel:1024 Loops:1024 Thr:1 Vec:8
```

One interesting thing that I stumbled upon during testing was that at some point I noticed that indeed with pocl got a half of the speed compared to the Intel. Hashcat reported around 400 MH/s with pocl and Intel around 800 MH/s consistently. Then I was able to alternate between 400 and 800 MH/s speed by compiling pocl just by chancing minor version number. When trying to figure this out I learned that hashcat actually has a kernel cache under home folder `.hashcat/kernels`. After deleting the cache I wasn't able to reproduce the lower speed anymore by trying out different versions of pocl and LLVM that gave me half of the speed before. Now all versions that I tried earlier were reporting around 800 MH/s consistently. So it was kernel caching issue but I've no idea what caused half of the speed kernel before removing the cache.

Few more important details that I remember during testing. This speed difference only happens with NTLM hash mode (-m1000). When using `export POCL_TRACING=cq` the slowest kernel was `m01000_s04` in both 400 and 800 MH/s situations but the average runtime in 400 case was around 16 ms and in 800 case around 9 ms. Also in 400 case `m01000_s04` kernel execution count was around half of the 800 case. These numbers also correlate with 400 and 800 speed differences. Sadly after removing the cache I wasn't able to replicate this anymore. So somehow hashcat or pocl ended up using half of the speed kernel which  explains the speed difference and this somehow relates to hashcat cache.

After this point I'm not using anymore of my time on this one. It cannot be said for sure is the issue with pocl or with hashcat.
